fix(replays): Updated replays list search input

danecando opened the pull request:
### Changes
- adds a new search input component that defines the search tags supported in replays

Closes #37968 

![Screen Shot 2022-08-31 at 2 54 52 PM](https://user-images.githubusercontent.com/3721977/187759999-dbc22fd6-e6b7-4e67-86d5-1e7babfb1bc8.png)


<!--

  Sentry employees and contractors can delete or ignore the following.

-->

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. and is gonna need some rights from me in order to utilize my contributions in this here PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

danecando committed:
fix(replays): omits unsupported tags from search

danecando committed:
fix(replays): uncomment unsupported fields

ryan953 commented on static/app/views/replays/filters.tsx:
```diff
@@ -47,3 +48,88 @@ const FilterContainer = styled('div')`
+  'transaction.duration',
+  'transaction.op',
+  'transaction.status',
+];
```
this list feels fragile, if someone adds something to the base list and we don't omit it then we'll have a problem.

Also, it looks like fields we should support are not part of the base list, so they're not in the typeahead at all

danecando committed:
fix(replays): use custom search input with defined tags

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,144 @@
+
+  return (
+    <SmartSearchBar
+      hasRecentSearches
```
I'm getting recent search results from discover queries, the fields are different I don't think it makes sense to show this right now
<img width="442" alt="Screen Shot 2022-09-01 at 1 31 55 PM" src="https://user-images.githubusercontent.com/187460/188006782-2282a704-863f-4ad6-bf78-d7748fbb91f1.png">

```suggestion
```

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,144 @@
+      }}
+      maxSearchItems={maxSearchItems}
+      excludeEnvironment
+      maxMenuHeight={maxMenuHeight ?? 300}
```
is this a reasonable default?

danecando replied:
I think so

<img width="539" alt="Screen Shot 2022-09-01 at 6 06 38 PM" src="https://user-images.githubusercontent.com/3721977/188021284-e8febf21-ad0c-4333-a73b-56d1a3be6ab3.png">

ryan953 commented on static/app/views/replays/filters.tsx:
```diff
@@ -2,13 +2,14 @@ import styled from '@emotion/styled';
 import ProjectPageFilter from 'sentry/components/projectPageFilter';
 import {t} from 'sentry/locale';
 import space from 'sentry/styles/space';
 import {Organization} from 'sentry/types';
 
+import SearchBar from './searchBar';
+
 type Props = {
```
can we use full imports. I find it much easier to grep for places that import a bit of code when all the callsites spell things the same way. I'm going to get the linter changed too

```suggestion
import {Organization} from 'sentry/types';
import SearchBar from 'sentry/views/replays/searchBar';

type Props = {
```

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,144 @@
+  'g'
+);
+
+const REPLAY_TAGS = [
+  'device.brand',
+  'device.family',
+  'device.name',
+  'dist',
+  'os.build',
+  'os.kernel_version',
+  'platform.name',
+  'release',
+  'sdk.name',
+  'sdk.version',
+  'user.display',
+  'user.email',
+  'user.id',
+  'user.ip',
+  'user.username',
+];
```
@cmanallen can you confirm the list of searchable fields?

danecando replied:
Some fields already exist in the regular API but they are named differently here. IE: ipAddress and the os fields. Should the API align on these or should I just add new search fields for these specifically for replay search?

ryan953 replied:
@cmanallen Can we double-check whether it's `user.ipAddress` (per your comment) or `user.ip_address` (api response returns this)


There are some missing fields in this list, which is fine. Listing them for clarity:
- title (idk what's in here, nvm this one)
- errorIds, traceIds, urls
- startedAt, finishedAt
- countUrls
- longestTransaction (gte search here would be cool)
- environment

danecando replied:
For the search it's ipAddress. If the search doesn't support the param it returns an error response. The reason I was asking is because existing search fields that we have already have descriptions and everything configured. So for these we'll need to create new ones specifically for replay queries (like I did with a few of the replay specific ones)

danecando committed:
fix(replays): feedback changes

danecando committed:
Merge remote-tracking branch 'upstream/master' into fix/replays/37968-replay-search

danecando committed:
fix(replays): adds replay specific search fields

ryan953 commented on static/app/utils/fields/index.ts:
```diff
@@ -432,16 +439,41 @@ export const SPAN_OP_FIELDS: Record<SpanOpBreakdown, FieldDefinition> = {
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.STRING,
+  },
+};
```
Looks like these can show up up in the Issues typeahead. I think it's because an error has a tag called 'duration' so we're pulling that description in. It's happening in the sentry org, not the test one.

The error i see with the duration tag is ae8e2a014c58452cb2c53f23e467e8fd

<img width="1079" alt="Screen Shot 2022-09-09 at 1 31 19 PM" src="https://user-images.githubusercontent.com/187460/189438183-8693e882-8ddd-4fef-ad63-cf0c363aa4e6.png">

danecando replied:
@ryan953 I'm not sure how I can add an item for the duration field and have it not show up in that view that is in this screenshot. Issue and performance searches have a list of defined fields and duration isn't one of them so for me it doesn't show up anywhere except the replay search. Currently I just added field items for the other fields that didn't have descriptions and left duration out. Maybe theres a search input for "My Pinned Search" where I can omit duration or I guess I would need to make changes to the underlying `SmartSearchBar` component

ryan953 replied:
It kinda feels like we need two versions of `static/app/utils/fields/index.ts`, and then inject the correct data into the searchBar instead of importing it. Otherwise with this big list of field names, we're going to have `duration: "The Duration of the Replay in ms"` as a key that'll match with anyones custom tags. The smartBar needs to only look at keys & descriptions which are related to it's domain, not everything under the sun.

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+  'id',
+  'os.build',
+  'os.kernel_version',
+  'platform.name',
```
```suggestion
  'platform',
```

per https://github.com/getsentry/sentry/pull/38343#discussion_r964879868

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+  'user.email',
+  'user.id',
+  'user.ip',
+  'user.username',
```
```suggestion
  'user.name',
```

per https://github.com/getsentry/sentry/pull/38343#discussion_r964879868

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+  'dist',
+  'duration',
+  'id',
+  'os.build',
+  'os.kernel_version',
```
```suggestion
  'os.name',
  'os.version',
```

per https://github.com/getsentry/sentry/pull/38343#discussion_r964879868

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+      maxSearchItems={maxSearchItems}
+      excludeEnvironment
+      maxMenuHeight={maxMenuHeight ?? 300}
+      {...props}
```
The order here looks slightly wrong. Someone could pass a prop `onGetTagValues` into this searchbar component and it'll override our `onGetTagValues={getEventFieldValues}` prop.

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+  // with data when ready
+  const getEventFieldValues = memoize(
+    (tag, query, endpointParams): Promise<string[]> => {
+      const projectIdStrings = (projectIds as Readonly<number>[])?.map(String);
```
The types here don't make sense to me. If we're forcing it to be a `Readonly<number>[]` then we don't need the optional-chaining `?` in there.

Shouldn't need to add anything extra imo:
```suggestion
      const projectIdStrings = projectIds?.map(String);
```

ryan953 commented on static/app/views/replays/searchBar.tsx:
```diff
@@ -0,0 +1,151 @@
+
+  const api = useApi();
+
+  useEffect(() => {
+    // Clear memoized data on mount to make tests more consistent.
+    getEventFieldValues.cache.clear?.();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [projectIds]);
```
We won't need this if `getEventFieldValues` is defined with `useCallback` instead of `memoize`. the dependencies for useCallback will be `api, organization.slug, projectIds`

danecando committed:
Merge remote-tracking branch 'upstream/master' into fix/replays/37968-replay-search

danecando committed:
fix(replays): review updates

danecando committed:
fix(replays): remove projectId from ReplayFieldKey

Removing any keys that aren't explicitly replay specific to avoid conflicts across the app

danecando commented:
I removed additional custom 'replay fields' if they aren't new fields that are specific to replays to avoid conflicts and also to just be more correct. Only downside is the ones that don't exist or have different keys are missing descriptions.

![Screen Shot 2022-09-12 at 12 02 21 PM](https://user-images.githubusercontent.com/3721977/189702012-94198770-abb2-4968-8cb1-b048d2952836.png)

I could add these alternate keyed fields IE: user.ipAddress and user.name along with the existing ones to provide descriptions. Let me know what you think about that (might just be a little confusing having duplicated fields defined)

I found a bug where if you visit a page with an existing search like the issues list and then go to replays and use the search field it gets loaded with all the tags from the issues search. I'm trying to figure this one out right now

danecando committed:
fix(replays): remove unused tags prop

ryan953 commented on static/app/utils/fields/index.ts:
```diff
@@ -170,6 +170,11 @@ export enum AggregationKey {
   LastSeen = 'last_seen',
 }
 
+export enum ReplayFieldKey {
```
doesn't look like the export is read by anything

danecando committed:
fix(replays): remove unused export

ryan953 commented on static/app/utils/fields/index.ts:
```diff
@@ -432,16 +437,31 @@ export const SPAN_OP_FIELDS: Record<SpanOpBreakdown, FieldDefinition> = {
   },
 };
 
+export const REPLAY_FIELDS: Record<ReplayFieldKey, FieldDefinition> = {
```
doesn't look like the export is read by anything

danecando committed:
fix(replays): remove unused export

danecando committed:
Merge remote-tracking branch 'upstream/master' into fix/replays/37968-replay-search

ryan953 commented:
> I removed additional custom 'replay fields' if they aren't new fields that are specific to replays to avoid conflicts and also to just be more correct. Only downside is the ones that don't exist or have different keys are missing descriptions.
> 
> ![Screen Shot 2022-09-12 at 12 02 21 PM](https://user-images.githubusercontent.com/3721977/189702012-94198770-abb2-4968-8cb1-b048d2952836.png)
> 
> I could add these alternate keyed fields IE: user.ipAddress and user.name along with the existing ones to provide descriptions. Let me know what you think about that (might just be a little confusing having duplicated fields defined)

I think all of our fields should have a description, but if it's a replay-specific field then it shouldn't showup in the auto-complete for events/discovery, and vis-versa. 

So like, in the screen shot `device.model` and `duration` both need descriptions, duration should probably include the unit that you get back

danecando committed:
fix(replays): add field items for searchable replay fields

danecando committed:
fix(replays): adds fields for browser name and version

danecando committed:
fix(replays): remove duration field
